File: lib/bit.go
// BIT (A Binary indexed tree or Fenwick tree)
type BIT struct {
	n int
	d []int
}

func NewBIT(n int, a ...int) *BIT {
	b := BIT{}
	b.n = n
	b.d = make([]int, n+1)
	for i := 0; i < len(a); i++ {
		b.Add(i, a[i])
	}
	return &b
}

func (b BIT) Add(i, x int) {
	for i++; i <= b.n; i += i & -i {
		b.d[i] += x
	}
}

func (b BIT) Sum(i int) int {
	x := 0
	for i++; i > 0; i -= i & -i {
		x += b.d[i]
	}
	return x
}

File: lib/rust/bit.rs
// BIT (A Binary indexed tree or Fenwick tree)
struct BIT {
    data: Vec<i32>,
}

impl BIT {
    fn new(n: usize) -> Self {
        Self {
            data: vec![0; n + 1],
        }
    }

    fn add(&mut self, mut i: usize, v: i32) {
        i += 1;
        while i < self.data.len() {
            self.data[i] += v;
            i += i & i.wrapping_neg();
        }
    }

    fn sum(&self, mut i: usize) -> i32 {
        let mut v = 0;
        while i > 0 {
            v += self.data[i];
            i -= i & i.wrapping_neg();
        }
        v
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sum() {
        let mut bit = BIT::new(3);
        for i in 0..3 {
            bit.add(i, i as i32 + 1);
        }
        assert!(bit.sum(0) == 0);
        assert!(bit.sum(1) == 1);
        assert!(bit.sum(2) == 3);
        assert!(bit.sum(3) == 6);
    }
}
